syntax = "proto3";

package user;

option go_package = "./pb";

// User service definition
service UserService {
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
  rpc GetUser(GetUserRequest) returns (GetUserResponse);
  rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse);
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
}

// User message
message User {
  string id = 1;
  string name = 2;
  string email = 3;
  int32 age = 4;
  int64 created_at = 5;
  int64 updated_at = 6;
}

// Create User
message CreateUserRequest {
  string name = 1;
  string email = 2;
  int32 age = 3;
}

message CreateUserResponse {
  User user = 1;
  string message = 2;
}

// Get User
message GetUserRequest {
  string id = 1;
}

message GetUserResponse {
  User user = 1;
  string message = 2;
}

// Update User
message UpdateUserRequest {
  string id = 1;
  string name = 2;
  string email = 3;
  int32 age = 4;
}

message UpdateUserResponse {
  User user = 1;
  string message = 2;
}

// Delete User
message DeleteUserRequest {
  string id = 1;
}

message DeleteUserResponse {
  string message = 1;
}

// List Users
message ListUsersRequest {
  int32 page = 1;
  int32 limit = 2;
}

message ListUsersResponse {
  repeated User users = 1;
  int32 total = 2;
  string message = 3;
}